openmdao/jacobians/assembled_jacobian.py

```diff
@@ -327,14 +333,90 @@ def _apply(self, d_inputs, d_outputs, d_residuals, mode):
         with system._unscaled_context(
                 outputs=[d_outputs], residuals=[d_residuals]):
             if mode == 'fwd':
-                d_residuals.iadd_data(int_mtx._prod(d_outputs.get_data(), mode, int_ranges))
-                if ext_mtx is not None:
-                    d_residuals.iadd_data(ext_mtx._prod(d_inputs.get_data(), mode, None))
+                if len(d_outputs._names) > 0 and len(d_residuals._names) > 0:
+
+                    d_residuals.iadd_data(int_mtx._prod(d_outputs.get_data(), mode, int_ranges))
+
+                if ext_mtx is not None and \
+                   len(d_inputs._names) > 0 and len(d_residuals._names) > 0:
+
+                    # Masking
+                    cache_key = tuple(d_inputs._names)
+                    if cache_key not in self._mask_caches:
+                        self._create_mask_cache(d_inputs, cache_key, ext_mtx)
+
+                    mask = self._mask_caches.get(cache_key)
+                    if mask is not None:
+                        inputs_masked = np.ma.array(d_inputs.get_data(), mask=mask)
+
+                        # Use the special dot product function from masking module so that we
+                        # ignore masked parts.
+                        d_residuals.iadd_data(np.ma.dot(ext_mtx._matrix, inputs_masked))
+
+                    else:
+                        d_residuals.iadd_data(ext_mtx._prod(d_inputs.get_data(), mode, None))
+
             else:  # rev
                 dresids = d_residuals.get_data()
-                d_outputs.iadd_data(int_mtx._prod(dresids, mode, int_ranges))
-                if ext_mtx is not None:
-                    d_inputs.iadd_data(ext_mtx._prod(dresids, mode, None))
+                if len(d_outputs._names) > 0 and len(d_residuals._names) > 0:
+
+                    d_outputs.iadd_data(int_mtx._prod(dresids, mode, int_ranges))
+
+                if ext_mtx is not None and \
+                   len(d_inputs._names) > 0 and len(d_residuals._names) > 0:
+
+                    # Masking
+                    cache_key = tuple(d_inputs._names)
+                    if cache_key not in self._mask_caches:
+                        self._create_mask_cache(d_inputs, cache_key, ext_mtx)
+
+                    mask_cols = self._mask_caches.get(cache_key)
+                    if mask_cols is not None:
+
+                        # Mask need to be applied to ext_mtx so that we can ignore multiplication
+                        # by certain columns.
+                        mask = np.zeros(ext_mtx._matrix.T.shape, dtype=np.bool)
+                        mask[mask_cols, :] = True
+                        masked_mtx = np.ma.array(ext_mtx._matrix, mask=mask, fill_value=0.0)
+
+                        masked_product = np.ma.multiply(masked_mtx.T, dresids).flatten()
+
+                        for set_name, data in iteritems(d_inputs._data):
+                            data += np.ma.filled(masked_product)
```

Comment:
It doesn't look like this would work with multiple varsets. If that's the case, we should put in a TODO to generalize this later for running with multiple varsets

Reply:
I will add a note for this and other advanced features.